dyno-recipes/src/main/java/com/netflix/dyno/recipes/lock/DynoLockClient.java

```diff
@@ -0,0 +1,369 @@
+        CountDownLatch latch = new CountDownLatch(quorum);
+        if (extend) {
+            votingHostsSelector.getVotingHosts().getEntireList().stream()
+                    .map(host -> new ExtendHost(host, pool, lockResource, latch, resourceKeyMap.get(resource)))
+                    .forEach(lH -> CompletableFuture.supplyAsync(lH, service));
+        } else {
+            votingHostsSelector.getVotingHosts().getEntireList().stream()
+                    .map(host -> new LockHost(host, pool, lockResource, latch, resourceKeyMap.get(resource)))
+                    .forEach(lH -> CompletableFuture.supplyAsync(lH, service));
+        }
+        awaitLatch(latch, resource);
+        long validity = 0L;
+        if (lockResource.getLocked() >= quorum) {
+            long timeElapsed = Instant.now().toEpochMilli() - startTime;
+            validity = ttlMS - timeElapsed - drift;
+        } else {
+            releaseLock(resource);
+        }
+        return validity;
+    }
+
+    /**
+     * Tries to acquire lock on resource for ttlMS milliseconds. Returns the amount of time for which the lock was acquired
+     * @param resource
+     * @param ttlMS
+     * @return
+     */
+    public long acquireLock(String resource, long ttlMS) {
+        resourceKeyMap.putIfAbsent(resource, getRandomString());
+        return runLockHost(resource, ttlMS, false);
+    }
+
+    boolean checkResourceExists(String resource) {
+        if (!resourceKeyMap.containsKey(resource)) {
+            logger.info("No lock held on {}", resource);
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    private boolean awaitLatch(CountDownLatch latch, String resource) {
+        try {
+            return latch.await(timeout, timeoutUnit);
+        } catch (InterruptedException e) {
+            logger.info("Interrupted while checking the lock for resource {}", resource);
+            return false;
+        }
+    }
+
+    /**
+     * Check and get the ttls for the lock if it exists. This returns the minimum of ttls returned across the hosts
+     * @param resource
+     * @return
+     */
+    public long checkLock(String resource) {
+        if (!checkResourceExists(resource)) {
+            return 0;
+        } else {
+            long startTime = Instant.now().toEpochMilli();
+            CopyOnWriteArrayList<Long> resultTtls = new CopyOnWriteArrayList<>();
+            CountDownLatch latch = new CountDownLatch(quorum);
+            votingHostsSelector.getVotingHosts().getEntireList().stream()
+                    .map(host -> new CheckAndRunHost(host, pool, "pttl", resource, resourceKeyMap.get(resource)))
+                    .forEach(checkAndRunHost -> CompletableFuture.supplyAsync(checkAndRunHost, service)
+                            .thenAccept(r -> {
+                                String result = r.getResult().toString();
+                                if (result.equals("0") || result.equals("-2")) {
```

Comment:
Comment about why it's 0 and -2. (Can also set a local const variable to represent both the values if you like)